test/versioned/grpc/util.js

```diff
@@ -0,0 +1,244 @@
+ *
+ * @param {Object} grpc grpc module
+ * @returns {Object} { server, proto } grpc server and protobuf api
+ */
+util.createServer = async function createServer(grpc) {
+  const server = new grpc.Server()
+  const credentials = grpc.ServerCredentials.createInsecure()
+  // quick and dirty map to store metadata for a given gRPC call
+  server.metadataMap = new Map()
+  const serverMethods = require('./grpc-server')(server)
+  const proto = loadProtobufApi(grpc)
+  server.addService(proto.Greeter.service, serverMethods)
+  await new Promise((resolve, reject) => {
+    server.bindAsync(SERVER_ADDR, credentials, (err, port) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(port)
+      }
+    })
+  })
+  server.start()
+  return { server, proto }
+}
+
+/**
+ * Gets the client for the Greeter service
+ *
+ * @param {Object} grpc grpc module
+ * @param {Object} proto protobuf API example.proto
+ * @returns {Object} client grpc client for Greeter service
+ */
+util.getClient = function getClient(grpc, proto) {
+  const credentials = grpc.credentials.createInsecure()
+  return new proto.Greeter(CLIENT_ADDR, credentials)
+}
+
+/**
+ * Gets the formatted substring for a given gRPC method
+ *
+ * @param {string} fnName name of gRPC call
+ * @returns {string}
+ */
+util.getRPCName = function getRPCName(fnName) {
+  return `/helloworld.Greeter/${fnName}`
+}
+
+/**
+ * Asserts the gRPC external segment and its relevant attributes: http.url,
+ * http.method, grpc.statusCode, grpc.statusText
+ *
+ * @param {Object} params
+ * @param {Object} params.t tap test
+ * @param {Object} params.tx transaction under test
+ * @param {string} params.fnName gRPC method name
+ * @param {number} [params.expectedStatusCode=0] expected status code for test
+ * @param {string} [params.expectedStatusText=OK] expected status text for test
+ */
+util.assertExternalSegment = function assertExternalSegment({
+  t,
+  tx,
+  fnName,
+  expectedStatusCode = 0,
+  expectedStatusText = 'OK'
+}) {
+  const methodName = util.getRPCName(fnName)
+  const segmentName = `${EXTERNAL.PREFIX}${CLIENT_ADDR}${methodName}`
+  metricsHelpers.assertSegments(tx.trace.root, [segmentName], { exact: false })
+  const segment = metricsHelpers.findSegment(tx.trace.root, segmentName)
+  const attributes = segment.getAttributes()
+  t.equal(
+    attributes['http.url'],
+    `grpc://${CLIENT_ADDR}${methodName}`,
+    'http.url attribute should be correct'
+  )
+  t.equal(attributes['http.method'], methodName, 'method name should be correct')
+  t.equal(
+    attributes['grpc.statusCode'],
+    expectedStatusCode,
+    `status code should be ${expectedStatusCode}`
+  )
+  t.equal(
+    attributes['grpc.statusText'],
+    expectedStatusText,
+    `status text should be ${expectedStatusText}`
+  )
+  t.equal(attributes.component, 'gRPC', 'should have the component set to "gRPC"')
+  metricsHelpers.assertMetrics(tx.metrics, [expectedMetrics], false, false)
+  t.end()
```

Comment:
Although convenient to have `t.end()` here, I get a little confused when reading the tests that call this function, because I am wondering, "where does the test end?", especially because we wrote the tests a little "upside-down", so to speak, with the ending portion of the test written first when the transaction ends and the attributes are asserted.

I would prefer to leave the `t.end()` inside the test bodies instead of this helper function, s'all I'm saying.

Reply:
I can change this but FWIW we have a lot of these style of tests in our repo